src/Data/Text/Internal/Reverse.hs

```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{-# OPTIONS_HADDOCK not-home #-}
+
+-- | Implements 'reverse', using efficient C routines by default.
+module Data.Text.Internal.Reverse (reverse) where
+
+#if !defined(PURE_HASKELL)
+import GHC.Exts as Exts
+import Control.Monad.ST.Unsafe (unsafeIOToST)
+import Foreign.C.Types (CSize(..))
+#else
+import Data.Word (Word8)
+import Control.Monad.ST (ST)
+#endif
+import Prelude hiding (reverse)
+import Data.Text.Internal (Text(..))
+import Control.Monad.ST (runST)
+import qualified Data.Text.Array as A
+
+-- | /O(n)/ Reverse the characters of a string.
+--
+-- Example:
+--
+-- >>> T.reverse "desrever"
+-- "reversed"
+reverse ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  Text -> Text
+#if defined(PURE_HASKELL)
+reverse (Text ba off len) = runST $ do
+    dest <- A.new len
+    let bytes = A.toList ba off len
```

Comment:
Could we avoid converting to a list? Why not iterate the array immediately?

Reply:
Yeah, I just grabbed the first thing that looked like it worked. I figured doing a reverse requires running `unsafeIndex` on each byte anyway, and I hoped any Word8 boxing might be avoided by GHC. But I haven't checked if that's the case.

Reply:
Reimplemented it with utf8LengthByLeader and copyI. New benchmark results are much nicer:

```
--- pure.result	2023-09-07 21:37:04.234194538 +0300
+++ pure-2.result	2023-09-10 22:07:52.516594503 +0300
@@ -9,40 +9,40 @@
   Pure
     tiny
       reverse
-        Text:     OK (0.38s)
-          173  ns ±  17 ns
-        LazyText: OK (0.21s)
-          200  ns ±  16 ns
+        Text:     OK (0.30s)
+          34.2 ns ± 2.1 ns
+        LazyText: OK (0.38s)
+          43.4 ns ± 1.3 ns
     ascii-small
       reverse
-        Text:     OK (0.13s)
-          2.11 ms ± 185 μs
-        LazyText: OK (0.13s)
-          2.15 ms ± 171 μs
+        Text:     OK (0.18s)
+          348  μs ±  30 μs
+        LazyText: OK (0.18s)
+          349  μs ±  32 μs
     ascii
       reverse
-        Text:     OK (4.05s)
-          1.294 s ±  39 ms
-        LazyText: OK (3.49s)
-          1.152 s ± 9.7 ms
+        Text:     OK (0.96s)
+          290  ms ±  19 ms
+        LazyText: OK (0.93s)
+          285  ms ± 3.9 ms
     english
       reverse
-        Text:     OK (0.57s)
-          76.7 ms ± 3.9 ms
-        LazyText: OK (0.24s)
-          76.8 ms ± 3.3 ms
+        Text:     OK (0.17s)
+          20.1 ms ± 1.4 ms
+        LazyText: OK (0.14s)
+          19.0 ms ± 1.4 ms
     russian
       reverse
-        Text:     OK (0.11s)
-          229  μs ±  22 μs
-        LazyText: OK (0.12s)
-          230  μs ±  22 μs
+        Text:     OK (0.18s)
+          44.0 μs ± 3.0 μs
+        LazyText: OK (0.18s)
+          44.0 μs ± 3.3 μs
     japanese
       reverse
-        Text:     OK (0.25s)
-          242  μs ±  20 μs
-        LazyText: OK (0.51s)
-          244  μs ± 5.7 μs
+        Text:     OK (0.12s)
+          29.1 μs ± 2.8 μs
+        LazyText: OK (0.88s)
+          25.7 μs ± 1.5 μs
 
-All 12 tests passed (11.46s)
+All 12 tests passed (5.37s)
 Benchmark text-benchmarks: FINISH
```